### PR TITLE
Show mlir program when tracing benchmarking

### PR DESCRIPTION
**Pull request by pfultz2**
This will show the mlir program when using `MIGRAPHX_TRACE_BENCHMARKING=3`.

### CODE REVIEW

File: src/targets/gpu/mlir.cpp [hunk 1]
```diff
@@ -960,6 +952,24 @@ void adjust_param_shapes(module& m, const std::vector<shape>& inputs)
...
 }
 
+std::string dump_mlir(const module& m, const std::vector<shape>& inputs)
+{
+    module mm;
+    const_module_ref mr = &m;
+    if(not inputs.empty())
+    {
+        mm = m;
+        mr = &mm;
```
Comment (umangyadav) [inline src/targets/gpu/mlir.cpp:962]: Why does it require const_ref ? Shouldn't just copied module `mm` work ?

Comment (pfultz2) [reply src/targets/gpu/mlir.cpp:962]: I only want to copy the module if there is no input shapes because we wont be adjusting the parameters.

Comment (umangyadav) [reply src/targets/gpu/mlir.cpp:962]: Module with empty input shape is unlikely case it would have been const-folded.

Comment (umangyadav) [reply src/targets/gpu/mlir.cpp:962]: I think it is better to assume inputs are not empty. it would be simpler

Comment (pfultz2) [reply src/targets/gpu/mlir.cpp:962]: There is two overloads to `dump_mlir`. One just takes the module(which we dont want to copy) and the other overload takes the input shapes, which might be different than the shapes in the module so we need to modify the module which we will use a copy for this case.

File: src/targets/gpu/compile_ops.cpp [hunk 2]
```diff
@@ -179,6 +185,8 @@ struct compile_plan
                                    std::cout << "No binary" << std::endl;
                                return std::numeric_limits<double>::max();
                            }
+                           if(trace_level > 2)
```
Comment (umangyadav) [inline src/targets/gpu/compile_ops.cpp:188]: add this option to documents

File: src/targets/gpu/mlir.cpp (hunk 1 above)
Comment (umangyadav) [reply src/targets/gpu/mlir.cpp:962]: > and the other overload takes the input shapes, which might be different than the shapes in the module so we need to modify the module which we will use a copy for this case. 

Yes but those input shapes parameter would be input arguments to the precompile_op instruction. If they are empty that means MLIR module also doesn't take any inputs.

Comment (pfultz2) [reply src/targets/gpu/mlir.cpp:982]: When the inputs are empty it means skip doing param adjustments.

File: src/targets/gpu/compile_ops.cpp [hunk 3]
```diff
@@ -153,22 +159,33 @@ struct compile_plan
...
+    {
+        if(config)
+            return to_string(config->problem);
+        return "<no problem key>";
+    }
+
     const compiled_result& benchmark() const
     {
         const auto trace_level = value_of(MIGRAPHX_TRACE_BENCHMARKING{});
+        if(trace_level > 0 and not results.empty())
```
Comment (bpickrel) [inline src/targets/gpu/compile_ops.cpp:172]: Suggest you consolidate some of these checks with `if ... or ...`

**Review by bpickrel — APPROVED** [review]

**Pull request merged**